web/cobrands/fixmystreet/assets.js

```diff
@@ -16,139 +16,91 @@ var fixmystreet = fixmystreet || {};
     },
 
-    check_for_road: function(lonlat) {
-        var road_providers = fixmystreet.map.getLayersBy('fixmystreet', {
-            test: function(options) {
-                return options && options.road && (options.all_categories || options.asset_category.indexOf($('select#form_category').val()) != -1);
-            }
-        });
-        if (road_providers.length) {
-            var road_layer = road_providers[0];
-            fixmystreet.roads.last_road = road_layer;
-            var point = new OpenLayers.Geometry.Point(lonlat.lon, lonlat.lat);
-            var feature = road_layer.getFeatureAtPoint(point);
-            if (feature == null) {
-                // The click wasn't directly over a road, try and find one
-                // nearby
-                feature = road_layer.getNearestFeature(point, 10);
-            }
-            if (feature !== null) {
-                selected_road = feature; //.attributes[road_layer.fixmystreet.road.attribute];
-            } else {
-                selected_road = null;
-            }
-            if (selected_road) {
-                fixmystreet.roads.found(road_layer, selected_road);
-            } else {
-                fixmystreet.roads.not_found(road_layer);
-            }
-        } else {
-            fixmystreet.roads.not_found();
-        }
+    one_time_select: function() {
+        // This function takes the current report lat/lon from hidden input
+        // fields and uses that to look up a USRN from the USRN layer.
+        // It's registered as an event handler by init_asset_layer below,
+        // and is only intended to run the once (because if the user drags the
+        // pin the usual USRN lookup event handler is run) so unregisters itself
+        // immediately.
+        this.events.unregister( 'loadend', this, this.one_time_select );
+        this.checkFeature(null, fixmystreet.get_lonlat_from_dom());
     },
 
-    found: function(layer, feature) {
-        if (layer.fixmystreet.actions) {
-            layer.fixmystreet.actions.found(layer, feature);
+    road_found: function() {
+        if (this.fixmystreet.actions) {
+            this.fixmystreet.actions.found(this, this.selected_feature);
         } else {
-            $('#single_body_only').val(layer.fixmystreet.body);
+            $('#single_body_only').val(this.fixmystreet.body);
         }
     },
 
-    not_found: function(layer) {
-        if (layer && layer.fixmystreet.actions) {
-            layer.fixmystreet.actions.not_found(layer);
+    road_not_found: function() {
+        if (this.fixmystreet.actions) {
+            this.fixmystreet.actions.not_found(this);
         } else {
-            if ( fixmystreet.roads.last_road && fixmystreet.roads.last_road.fixmystreet.actions.unselected ) {
-                fixmystreet.roads.last_road.fixmystreet.actions.unselected();
-                fixmystreet.roads.last_road = null;
+            if ( this.fixmystreet.actions.unselected ) {
```

Comment:
If there is no `this.fixmystreet.actions` then surely this will never be true?

Reply:
Hmm, yes. I admit I didn't really understand this. `not_found` is either called right after `last_road` was set (and so no need to look it up), or if the road layer had disappeared, and I wasn't sure how that could happen. It doesn't look like `unselected` would ever be called?

Reply:
The purpose is that if you need to do cleanup when a feature is unselected (e.g. hiding some additional text displayed) it is run. Previously this worked by storing the last selected layer and running unselected on that in not_found. In this case we shouldn't need to do that as each layer recieves the event so can call it's own unselected function.